src/onnx/parse_softmaxcrossentropyloss.cpp

```diff
@@ -0,0 +1,450 @@
+        }
+        else if(reduction == "mean" and not has_weights)
+        {
+            final_loss_tensor = info.add_instruction(migraphx::make_op("reduce_mean", {{"axes", loss_dims}}),
+                                               final_loss_tensor);
+        }
+        else if(reduction == "sum")
+        {
+            final_loss_tensor =
+                info.add_instruction(migraphx::make_op("reduce_sum", {{"axes", loss_dims}}), final_loss_tensor);
+        }
+
+        return final_loss_tensor;
+    }
+
+    std::vector<instruction_ref> parse(const op_desc& /*opd */,
+                                       const onnx_parser& parser,
+                                       const onnx_parser::node_info& info,
+                                       const std::vector<instruction_ref>& args) const
+    {
+        // Get and handle attributes
+        auto reduction = get_reduction_param(info);
+
+        // Get and validate Inputs
+        auto scores = get_scores(args.at(0));
+        auto scores_shape = scores->get_shape();
+        auto labels  = get_labels(args.at(1), scores_shape);
+
+        // meta parameters based on input scores shape
+        size_t ndims = scores_shape.ndim();
+        size_t batch_size = scores_shape.lens().at(0);
+        size_t class_size = scores_shape.lens().at(1);
+        bool is_k_dim  = (ndims > 3);
+        bool sizes_mismatch = batch_size != class_size;
+
+        // ignore_index is optional attribute, assign this as a scalar literal input to the op
+        instruction_ref ignore_index;
+        auto has_ignore_index =
+            normalize_input_index(parser, info, static_cast<int64_t>(class_size), ignore_index);
+
+        bool has_weights = (args.size() > 2);
+        auto weights = get_weights(info, args, ignore_index, scores_shape, class_size, has_ignore_index);
+
+        // Need to perform softmax on all the data before we select final axes
+        scores =
+            info.add_instruction(migraphx::make_op("softmax", {{"axis", 1}}), scores);
+
+        // This can be skipped if batch dim == class dim and there's only one batch dimension
+        // The entire op then becomes a much simpler mul + gather + scatter + reduce after pointwise operations
+        if(is_k_dim or sizes_mismatch) 
+        {
+            auto gathernd_indicies = handle_index_selection(info, labels);
+
+            std::vector<int64_t> perm(class_size, 0);
+            if (is_k_dim)
+            {
+                std::iota(++perm.begin(), perm.end(), 2);
+                perm.at(class_size - 1) = 1;
+                scores = info.add_instruction(migraphx::make_op("transpose", {{"permutation", perm}}), scores);
+            }
+
+            scores = info.add_instruction(migraphx::make_op("gathernd"), scores, gathernd_indicies);
+
+            if (has_weights) // Saves us a multiply + gather op to gate this as default weights are literal 1's
```

Comment:
See above comment, and also you dont have to worry about this optimization in the parser. There is a matcher that looks for a multiplication with a literal of all 1s. That will take care of it.
I suggest getting rid of all the code branches that check for this to make the parser easier to follow.

Reply:
All good. removed